### PR TITLE
feat: [CDS-69391]: Merge RunPipelineForm and RetryPipeline

### PR DESCRIPTION
**Pull request by srdjan-harness**
### Summary

Implementation is done by applying diff from RetryPipeline to RunPipelineForm.

To distinguish if its "retry from failed stage" RunPipelineForm has two new props:   
  isRetryFromStage,
  preSelectLastStage 
if isRetryFromStage is true, RunPipelineForm form will behave like RetryPipeline.

Added SelectStageToRetryNew (controlled) components that has all logic for selection stage/s to start execution from.

FF added:  CDS_MERGED_RUN_AND_RETRY_PIPELINE_COMPONENT

Doc created for this implementation:
https://harness.atlassian.net/wiki/spaces/~6290ab021a437e0070443d18/pages/21451638058/Refactor+Run+Pipeline+Form+Retry+Pipeline+Form+into+one+UI+component.

Jira:
https://harness.atlassian.net/browse/CDS-69391

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by srdjan-harness** [comment]
retrigger eslint

**Comment by srdjan-harness** [comment]
retrigger titlecheck

**Comment by srdjan-harness** [comment]
retrigger titlecheck

**Comment by srdjan-harness** [comment]
retrigger prittier

**Comment by srdjan-harness** [comment]
retrigger titlecheck

**Comment by srdjan-harness** [comment]
retrigger titlecheck

File: src/modules/70-pipeline/components/ExecutionActions/ExecutionActions.tsx [hunk 1]
```diff
@@ -366,7 +373,7 @@ const ExecutionActions: React.FC<ExecutionActionsProps> = props => {
                 data-test-id="retry-failed-pipeline"
                 featuresProps={getFeaturePropsForRunPipelineButton({ modules, getString })}
                 text={getString('pipeline.execution.actions.reRunFromStage')}
-                onClick={retryPipeline}
+                onClick={() => retryPipeline()}
```
Comment (rameshx) [inline src/modules/70-pipeline/components/ExecutionActions/ExecutionActions.tsx:376]: Is this needed?

File: src/modules/70-pipeline/components/RunPipelineModal/RunPipelineForm.module.scss [hunk 2]
```diff
@@ -307,3 +307,12 @@
   width: 600px;
   padding-top: 48px !important;
 }
+
+.selectStageWrapper {
+  padding: 24px;
+  border-bottom: 0.5px solid rgb(217, 218, 230);
```
Comment (rameshx) [inline src/modules/70-pipeline/components/RunPipelineModal/RunPipelineForm.module.scss:313]: let's use the existing variable `--grey-200`

File: src/modules/70-pipeline/components/RunPipelineModal/RunPipelineForm.tsx [hunk 3]
```diff
@@ -423,6 +478,44 @@ function RunPipelineFormBasic({
...
+  })
+
+  const retryStagesResponseData = retryStagesResponse?.data
+
+  useEffect(() => {
+    if (isRetryFromStage) {
+      fetchRetryStages()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isRetryFromStage])
```
Comment (rameshx) [inline src/modules/70-pipeline/components/RunPipelineModal/RunPipelineForm.tsx:510]: The boolean `isRetryFromStage` can probably be used for `lazy` which should help in removing this effect.

File: src/modules/70-pipeline/components/RunPipelineModal/SelectStageToRetryNew.tsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+      isAllStage,
+      isParallelStage,
+      listOfSelectedStages,
+      selectedStage
+    }
+    if (e.currentTarget.value === 'allparallel') {
+      newState.isAllStage = true
+    } else {
+      newState.isAllStage = false
+    }
```
Comment (rameshx) [inline src/modules/70-pipeline/components/RunPipelineModal/SelectStageToRetryNew.tsx:176]:
```suggestion
     newState.isAllStage = e.currentTarget.value === 'allparallel'
```

File: src/modules/70-pipeline/components/RunPipelineModal/__tests__/RunPipelineForm.test.tsx [hunk 5]
```diff
@@ -756,3 +763,194 @@ describe('EXECUTION VIEW', () => {
...
+
+  test('Retry Failed Pipeline title and button to be defined', () => {
+    const { queryAllByText } = render(
+      <TestWrapper>
+        <RunPipelineForm {...commonProps} source="executions" isRetryFromStage={true} />
+      </TestWrapper>
+    )
+    expect(queryAllByText('pipeline.execution.actions.reRun')[0]).not.toBeNull()
+    expect(queryAllByText('pipeline.execution.actions.reRunSpecificStageTitle').length).toEqual(1)
+    expect(queryAllByText('pipeline.execution.actions.reRun').length).toEqual(1)
```
Comment (rameshx) [inline src/modules/70-pipeline/components/RunPipelineModal/__tests__/RunPipelineForm.test.tsx:780]: Nit: Let's use `queryBy*` for checking the absence of elements and `getBy*` or `findBy*` for checking the presence of elements.

**Review by rameshx — APPROVED** [review]

**Review by vikrant-harness — DISMISSED** [review]

**Review by PratyushGar — DISMISSED** [review]

**Review by rameshx — APPROVED** [review]

**Comment by srdjan-harness** [comment]
retrigger cypress-cv

File: src/modules/70-pipeline/components/RunPipelineModal/__tests__/RunPipelineForm.test.tsx [hunk 6]
```diff
@@ -756,3 +764,194 @@ describe('EXECUTION VIEW', () => {
...
+
+    // Submit the incomplete form
+    const runButton = container.querySelector('button[type="submit"]')
+    act(() => {
+      fireEvent.click(runButton!)
+    })
+
+    const buttonShouldBeDisabled = container.querySelector('.bp3-disabled')
+    expect(buttonShouldBeDisabled).toBeTruthy()
+    expect(container).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (chetanvekariya-harness) [reply src/modules/70-pipeline/components/RunPipelineModal/__tests__/RunPipelineForm.test.tsx:956]: +1

**Review by chetanvekariya-harness — APPROVED** [review]

**Pull request merged**